File: solutions/python/maze_game.py
# @include
from abc import ABC, abstractmethod

class Room(ABC):

  @abstractmethod
  def connect(self, room2):
    pass

class MazeGame(ABC):

  @abstractmethod
  def make_room(self):
    print("abstract make_room")
    pass

  def addRoom(self, room):
    print("adding room")

  def __init__(self):
    room1 = self.make_room()
    room2 = self.make_room()
    room1.connect(room2)
    self.addRoom(room1)
    self.addRoom(room2)
# @exclude

File: solutions/java/MazeGame.java
// @include
public abstract class MazeGame {
  // Note the use of the template method pattern - subclasses
  // implement MakeRoom() as appropriate for the type of room being created.
  public MazeGame() {
    Room room1 = makeRoom();
    Room room2 = makeRoom();
    room1.connect(room2);
    this.addRoom(room1);
    this.addRoom(room2);
  }

  // Implementing classes provide this method.
  abstract protected Room makeRoom();
  // @exclude
  private void addRoom(Room that) {}
  // @include
}
// @exclude

interface Room {
  public void connect(Room that);
}
